### DISCUSSION TITLE
Realmに配列を保存したい

### DISCUSSION THREAD

**Discussion by mahirooo** [lit-iphone-course/lit-iphone-course#61] [Question-original] [answered]
(1) 実現したいこと
Realm上にstampImage配列を保存したい

(2) 実現するために自分が試した内容とその結果
①RealmにString型等を保存するときと同様にコードを書いた
→`Thread 1: "Property Stamp.stamp is declared as Array<String>, which is not a supported managed Object property type. If it is not supposed to be a managed property, either add it to `ignoredProperties()` or do not declare it as `@objc dynamic`. See https://www.mongodb.com/docs/realm-sdks/swift/latest/Classes/Object.html for more information."`
というエラーが出てどうやらRealmに配列は追加できなさそう？と気づく

②[このサイト](https://qiita.com/rei_012/items/df430d228577c041d61d)を参考に、配列を保存しようとしたが、stampImage配列をどのように保存すればよいのか分からなくなってしまった

(3) 調べた内容や資料
https://teratail.com/questions/104017
https://qiita.com/rei_012/items/df430d228577c041d61d

(4) 原因だと思われそうな箇所(ソースコードをコピペして貼る)
```
let realm = try Realm()
                       let dictionary: [String: Any] =
                           ["stamps": [["ticketTitle": "算数"],
                                        ["ticketTitle": "英語"],
                                        ["ticketTitle": "社会"]]
                           ]
                       let task = Stamp(value: dictionary) //Taskモデルのインスタンスの作成

                       //書き込み処理
                       try! realm.write {
                           realm.add(task)
                           print(task)
```

[Gitコードリンク](https://github.com/mahirooo/TampStamp/blob/df3e678d6fdf5192aafef9484b7bbd3caa416dcc/TampStamp/PushViewController.swift#L59-L79)

```
class Stamp: Object {
    @objc dynamic var stamp: String = ""
    //Listの定義
    let stampArray = List<StampArray>()
}

class StampArray: Object {
    @objc dynamic var ticketTitle: String = ""
}
```

[Gitコードリンク](https://github.com/mahirooo/TampStamp/blob/aedf9def571199e3e9cbb424fd02a40b97a9d18f/TampStamp/RealmSave.swift#L18-L26)

(5) GitHubのリンク
https://github.com/mahirooo/TampStamp

**Comment by yukiny0811** [comment, marked as answer]
こういうデータを定義してる時、
```.swift
class Parent: Object {
    @objc dynamic var title: String?
    let children = List<Child>()
}
class Child: Object {
    @objc dynamic var title: String?
}
```
Parentの中のchildrenに新しく要素を足すためには下みたいなコードを書かなきゃいけない！
Parentのchildrenの配列をいったん取ってきて、そこにChild()を追加する処理を```realm.write```の中で書く！
```.swift
let parents: Results<Parent> = realm.objects(Parent.self) //ここの2行
let parent = parents[0] //ここの2行は必要に応じて変えてね

let children = parent.children
let child = Child()
child.title = "aaaaa"
try! realm.write(){
    children.append(child)
}
```

**Reply by yukiny0811** [reply]
もし新しくParentを作って、そこにchildをたくさん足したいなら、
```.swift
let parent = Parent()
let children = parent.children

let child1 = Child()
child1.title = "aaaaa"

let child2 = Child()
child2.title = "bbbbb"

children.append(child1)
children.append(child2)

try! realm.write(){
    realm.add(parent)
}
```
って感じにしなきゃいけない！

**Reply by yukiny0811** [reply]
これ参考になるかも！
https://github.com/yukiny0811/realm-releation-test

**Reply by mahirooo** [reply]
ありがとううう！
上記参考に、PushViewControllerの画面を長押ししたときに、child.titleにstampImage配列の中身を全て追加されるようにコードを書いたら以下のエラーが出てしまったので、お手隙の時に見てほしいです👀

Build Succeed→実機が立ち上がって真っ白な画面のままエラー
`Thread 1: "Property Parent.title is declared as Array<String>, which is not a supported managed Object property type. If it is not supposed to be a managed property, either add it to `ignoredProperties()` or do not declare it as `@objc dynamic`. See https://www.mongodb.com/docs/realm-sdks/swift/latest/Classes/Object.html for more information."
`


Githubにもプッシュしてます！

**Reply by yukiny0811** [reply]
<img width="394" alt="image" src="https://user-images.githubusercontent.com/28947703/192503579-4f395196-c442-438b-ad27-15a9fb979375.png">
Realmのモデル定義のところでは[配列]は一切使えなくて、もし配列をそのまま保存したいなら、

```.swift
class Parent: Object {
    let children = List<Child>()
}
class Child: Object {
    @objc dynamic var title: String = ""
}
```
こういうモデルを作っておいて、
stampImageの配列[aaa, bbb, ccc, ddd, eee]の要素をそれぞれChildに変換しなきゃいけない！

具体的には
```.swift
let parent = Parent()
let children = parent.children
for img in stampImage {
    let child = Child()
    child.title = img
    children.append(child)
}
try! realm.write() {
    realm.add(parent)
}
```
こんな感じ！

**Reply by mahirooo** [reply]
あああなるほど、、配列を小分けにして保存するイメージで合ってるかな...？
ちょっとやってみます！

**Reply by yukiny0811** [reply]
そうそう！小分けにするイメージであってる！

**Reply by mahirooo** [reply]
保存できた！！ただ別の疑問が出てきたので別スレで質問します！

**Reply by yukiny0811** [reply]
OKー！
